go-controller/pkg/node/secondary_node_network_controller.go

```diff
@@ -122,11 +122,108 @@ func (oc *SecondaryNodeNetworkController) getNetworkID() (int, error) {
+		for _, udnSubnet := range oc.Subnets() {
+			if err := insertIptRules(getUDNSubnetReturnRule(udnSubnet.CIDR,
+				getIPTablesProtocol(udnSubnet.CIDR.IP.String()))); err != nil {
+				return err
+			}
+		}
+	} else {
+		for _, udnSubnet := range oc.Subnets() {
+			if err := deleteIptRules(getUDNSubnetReturnRule(udnSubnet.CIDR,
+				getIPTablesProtocol(udnSubnet.CIDR.IP.String()))); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// updateUDNVRFIPRule modifies existing IP rule for an UDN nework when that particular
+// network(10.132.0.0/14) is advertised:
+// Existing IP rule after creation of UDN network:
+// 2000:	from all fwmark 0x1001 lookup 1009
+// 2000:	from all to 169.254.0.12 lookup 1009
+// IP rule after the network is advertised:
+// 2000:	from all fwmark 0x1001 lookup 1009
+// 2000:	from all to 10.132.0.0/14 lookup 1009
+func (oc *SecondaryNodeNetworkController) updateUDNVRFIPRule() error {
+	interfaceName := util.GetNetworkScopedK8sMgmtHostIntfName(uint(oc.gateway.networkID))
+	mplink, err := util.LinkByName(interfaceName)
+	if err != nil {
+		return fmt.Errorf("unable to get link for %s, error: %v", interfaceName, err)
+	}
+	vrfTableId := util.CalculateRouteTableID(mplink.Attrs().Index)
+
+	if err = oc.gateway.ruleManager.DeleteWithMetadata(oc.gateway.GetNetworkRuleMetadata()); err != nil {
+		return fmt.Errorf("unable to delete iprule for network %s, err: %v", oc.gateway.GetNetworkName(), err)
+	}
+	udnReplyIPRules, err := oc.gateway.constructUDNVRFIPRules(vrfTableId)
+	if err != nil {
+		return fmt.Errorf("unable to get iprules for network %s, err: %v", oc.gateway.GetNetworkName(), err)
+	}
+	for _, rule := range udnReplyIPRules {
+		if err = oc.gateway.ruleManager.AddWithMetadata(rule, oc.gateway.GetNetworkRuleMetadata()); err != nil {
+			return fmt.Errorf("unable to create iprule %v for network %s, err: %v", rule, oc.gateway.GetNetworkName(), err)
+		}
+	}
+	return nil
+}
+
+// updateUDNFlow adds below OpenFlows based on the gateway mode and whether the network
+// is advertised or not:
+// table=1, n_packets=0, n_bytes=0, priority=16,ip,nw_dst=128.192.0.2 actions=LOCAL (Both gateway modes)
+// table=1, n_packets=0, n_bytes=0, priority=15,ip,nw_dst=128.192.0.0/14 actions=output:3 (shared gateway mode)
+func (oc *SecondaryNodeNetworkController) updateUDNFlow() error {
+	node, err := oc.gateway.watchFactory.GetNode(oc.gateway.nodeIPManager.nodeName)
+	if err != nil {
+		return fmt.Errorf("unable to get node %s: %s", oc.gateway.nodeIPManager.nodeName, err)
+	}
+	subnets, err := util.ParseNodeHostSubnetAnnotation(node, types.DefaultNetworkName)
+	if err != nil {
+		return fmt.Errorf("failed to get subnets for node: %s for OpenFlow cache update; err: %w", node.Name, err)
+	}
+	if err := oc.gateway.openflowManager.updateBridgeFlowCache(subnets, oc.gateway.nodeIPManager.ListAddresses(),
+		oc.gateway.isPodNetworkAdvertised, oc.gateway.isUDNNetworkAdvertised); err != nil {
+		return err
+	}
 	return nil
 }
```

Comment:
I would move these two methods to `UserDefinedNetworkGateway` as it is responsible for what it is being done (and undone) here.

I would also have them under a general `UserDefinedNetworkGateway.Reconcile` method just as we have `UserDefinedNetworkGateway.gateway.Reconcile` which you can reuse from the former to sync flows, and you can also reuse `UserDefinedNetworkGateway.gateway.Reconcile`